cub/cub/agent/agent_topk.cuh

```diff
@@ -0,0 +1,910 @@
+          // true, we need to write to `out` since it's the last chance.
+          else if ((out_buf || early_stop) && previous_bits < kth_key_bits)
+          {
+            NumItemsT pos   = atomicAdd(p_out_cnt, static_cast<NumItemsT>(1));
+            d_keys_out[pos] = key;
+            _CCCL_IF_CONSTEXPR (!KEYS_ONLY)
+            {
+              NumItemsT index   = in_idx_buf ? in_idx_buf[i] : i;
+              d_values_out[pos] = d_values_in[index];
+            }
+          }
+        };
+      if (load_from_ori_input)
+      {
+#ifdef USE_CUSTOMIZED_LOAD
+        VectorizedProcess(
+          static_cast<size_t>(blockIdx.x) * blockDim.x + threadIdx.x,
+          static_cast<size_t>(blockDim.x) * gridDim.x,
+          d_keys_in,
+          previous_len,
+          f);
+#else
+        ConsumeRange(d_keys_in, previous_len, f);
+#endif
+      }
+      else
+      {
+#ifdef USE_CUSTOMIZED_LOAD
+        VectorizedProcess(
+          static_cast<size_t>(blockIdx.x) * blockDim.x + threadIdx.x,
+          static_cast<size_t>(blockDim.x) * gridDim.x,
+          in_buf,
+          previous_len,
+          f);
+#else
+        ConsumeRange(in_buf, previous_len, f);
+#endif
+      }
+    }
+    if (early_stop)
+    {
+      return;
+    }
+    __syncthreads();
+
+    // merge histograms produced by individual blocks
+    for (int i = threadIdx.x; i < num_buckets; i += blockDim.x)
+    {
+      if (histogram_smem[i] != 0)
+      {
+        atomicAdd(histogram + i, histogram_smem[i]);
+      }
+    }
+  }
+
+  /**
+   * Replace histogram with its own prefix sum
+   * (step 2 in `radix_kernel` description)
+   */
+
+  _CCCL_DEVICE _CCCL_FORCEINLINE void Scan(volatile NumItemsT* histogram)
+  {
+    if (num_buckets >= BLOCK_THREADS)
+    {
+      // static_assert(num_buckets % BLOCK_THREADS == 0); //@todo: release constraints
+      constexpr int items_per_thread = (num_buckets - 1) / BLOCK_THREADS + 1;
+      typedef cub::BlockLoad<NumItemsT, BLOCK_THREADS, items_per_thread, cub::BLOCK_LOAD_TRANSPOSE> BlockLoad;
+      typedef cub::BlockStore<NumItemsT, BLOCK_THREADS, items_per_thread, cub::BLOCK_STORE_TRANSPOSE> BlockStore;
+      typedef cub::BlockScan<NumItemsT, BLOCK_THREADS> BlockScan;
+
+      __shared__ union
+      {
+        typename BlockLoad::TempStorage load;
+        typename BlockScan::TempStorage scan;
+        typename BlockStore::TempStorage store;
+      } temp_storage;
+      NumItemsT thread_data[items_per_thread];
+
+      BlockLoad(temp_storage.load).Load(histogram, thread_data, num_buckets, 0);
+      __syncthreads();
+
+      BlockScan(temp_storage.scan).InclusiveSum(thread_data, thread_data);
+      __syncthreads();
+
+      BlockStore(temp_storage.store).Store(histogram, thread_data, num_buckets);
+    }
+    else
+    {
+      typedef cub::BlockScan<NumItemsT, BLOCK_THREADS> BlockScan;
+      __shared__ typename BlockScan::TempStorage temp_storage;
```

Comment:
Totally agree, related code has been added https://github.com/ChristinaZ/cccl/blob/christinaz/nondeterministic_AIR_topK/cub/cub/agent/agent_topk.cuh#L235 
Besides that, I change the number of blocks at https://github.com/ChristinaZ/cccl/blob/christinaz/nondeterministic_AIR_topK/cub/cub/device/dispatch/dispatch_topk.cuh#L467, so that we can always launch at most 1 wave. And changed the usage of blockhead at https://github.com/ChristinaZ/cccl/blob/christinaz/nondeterministic_AIR_topK/cub/cub/agent/agent_topk.cuh#L421